kafka-connect/src/main/java/io/tabular/iceberg/connect/channel/Deduplicated.java

```diff
@@ -0,0 +1,254 @@
+
+import static java.util.stream.Collectors.toList;
+
+import io.tabular.iceberg.connect.events.CommitResponsePayload;
+import io.tabular.iceberg.connect.events.Event;
+import io.tabular.iceberg.connect.events.EventType;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class Deduplicated {
+  private static final Logger LOG = LoggerFactory.getLogger(Deduplicated.class);
+  private final UUID currentCommitId;
+  private final TableIdentifier tableIdentifier;
+  private final List<Envelope> deduplicatedEnvelopes;
+
+  private static final String DATA_FILE_TYPE = "data";
+  private static final String DELETE_FILE_TYPE = "delete";
+
+  Deduplicated(UUID currentCommitId, TableIdentifier tableIdentifier, List<Envelope> batch) {
+    this.currentCommitId = currentCommitId;
+    this.tableIdentifier = tableIdentifier;
+    this.deduplicatedEnvelopes = deduplicateEnvelopes(ImmutableList.copyOf(batch));
+  }
+
+  /**
+   * Returns the deduplicated data files from the batch of envelopes. Does not guarantee anything
+   * about the ordering of the files that are returned.
+   */
+  public List<DataFile> dataFiles() {
+    List<DataFile> dedupedInsidePayloads =
+        deduplicatedEnvelopes.stream()
+            .flatMap(
+                envelope -> {
+                  Event event = envelope.event();
+                  CommitResponsePayload payload = (CommitResponsePayload) event.payload();
+                  List<DataFile> dataFiles = payload.dataFiles();
+                  if (dataFiles == null) {
+                    return Stream.empty();
+                  } else {
+                    return deduplicateFilesInPayload(
+                        dataFiles,
+                        Deduplicated::dataFilePath,
+                        DATA_FILE_TYPE,
+                        payload.commitId(),
+                        envelope.partition(),
+                        envelope.offset(),
+                        event.id(),
+                        event.groupId(),
+                        event.type(),
+                        event.timestamp())
+                        .stream();
+                  }
+                })
+            .collect(toList());
+
+    List<DataFile> dedupedInsideAndAcrossPayloads =
+        deduplicateFiles(dedupedInsidePayloads, Deduplicated::dataFilePath, DATA_FILE_TYPE);
+
+    return ImmutableList.copyOf(dedupedInsideAndAcrossPayloads);
```

Comment:
Note that I could deduplicate a lot more easily in just a single pass. 

The only reason I'm doing 3 passes (once over envelopes, once over each payload, and once over all payloads) is so that if there are any duplicates, we get a log message telling us exactly which one of those cases we hit and hopefully that will lead us closer to a root cause.

Reply:
I would rather not lose logging details for this, I think doing the 3 passes as done here is good!